### PR TITLE
PEP 611: 'The one million limit'

### PR DESCRIPTION
**Pull request by markshannon**
Draft version of PEP 611, "The one million limit".

### CODE REVIEW

File: pep-0611.rst [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+A limit of one million fits into 20 bits which, although not as convenient for machine representation,
+is still reasonably compact. Three signed valuses in the range -1000_000 to +1000_000 can fit into a 64 bit word.
+A limit of one million is small enough for efficiency advantages (only 20 bits),
+but large enough not to impact users (no one has ever written a module of one million lines).
+
+The value "one million" is very easy to remember.
+
+The one million limit is mostly a limit on human generated code, not runtime sizes.
+
+One million lines in a single module is a rediculous concentration of code;
```
Comment (JelleZijlstra) [inline pep-0611.rst]: ridiculous

File: pep-0611.rst [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+It is unsafe as malicious or poorly generated code could cause values to exceed 2\ :sup:`32`.
+
+For example, line numbers are represented by 32 bit values internally. 
+This is inefficient, given that modules almost never exceed a few thousand lines.
+Despite being inefficent, is is still vulnerable to overflow as
+it is easy for an attacker to created a module with billions of newline characters.
+
+Memory access is usually a limiting factor in the performance of modern CPUs.
+Better packing of data structures enhances locality and reduces memory bandwith,
+at a modest increase in ALU usage (for shifting and masking).
```
Comment (RMJ10) [inline pep-0611.rst]: I don't think this assertion holds much water:
1. Caching make memory access much less of a limit than you would expect.
2. Non-aligned memory access vary from inefficient to impossible depending on the processor.
3. Shifting and masking isn't free, and again on some processors can be very expensive.

File: pep-0611.rst [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+Being able to safely store important values in 20 bits would allow memory savings
+in several data structures including, but not limited to:
+
+* Frame objects
+* Object headers
+* Code objects
+
+There is also the potential for a more efficient instruction format, speeding up interpreter dispatch.
```
Comment (RMJ10) [inline pep-0611.rst]: This is the ARM/IBM mistake all over again.

**Comment by Rosuav** [comment]
Folks, can we refrain from using this PR as a vehicle for discussions about the *contents* of the PEP please? Minor typos perhaps ("ridiculous"), but let's just get the PEP into the repo first, and then have the discussion on python-ideas.

@markshannon, Travis is showing a problem with the markup. Once that's sorted, this can be merged.

File: pep-0611.rst (hunk 1 above)
Comment (brettcannon) [inline pep-0611.rst]:
```suggestion
One million lines in a single module is a ridiculous concentration of code;
```

**Comment by RMJ10** [comment]
Apologies all, I confused myself into believing discussion had been redirected here.  This is what I get for following the link to the PR while updating a wiki for another project and writing JIRA tickets for yet another.  Clearly I need a lower limit on Things Rhodri Is Supposed To Do Simultaneously :-)

**Comment by Rosuav** [comment]
Clearly that number should be capped at one million :-)

File: pep-0611.rst (hunk 1 above)
Comment (markshannon) [reply pep-0611.rst]: thanks

File: pep-0611.rst (hunk 3 above)
Comment (markshannon) [reply pep-0611.rst]: Could you elaborate? Please bear in mind that this is software dispatching and decoding, not hardware.

File: pep-0611.rst (hunk 1 above)
Comment (markshannon) [reply pep-0611.rst]: thanks

**Comment by Rosuav** [comment]
Merged. Should be live soon. Let the bikeshedding commence!!

Oh wait, it already has. Have fun! :)

**Pull request merged**